tools/ot/src/main/java/dk/alexandra/fresco/tools/ot/base/ECNaorPinkasOt.java

```diff
@@ -0,0 +1,126 @@
+    super(otherId, randBit, network);
+    Security.addProvider(new BouncyCastleProvider());
+    X9ECParameters ecP = CustomNamedCurves.getByName("curve25519");
+    this.curve = ecP.getCurve();
+    this.subgroupOrder = curve.getOrder();
+    this.generator = ecP.getG();
+  }
+
+  @Override
+  BouncyCastleECCElement generateRandomNaorPinkasElement() {
+    // Sample random integer over the field of the curve
+    BigInteger randomInt = this.randNum.nextBigInteger(curve.getField().getCharacteristic());
+    // Compute the closets point on the curve to this
+    return new BouncyCastleECCElement(computePoint(randomInt));
+  }
+
+  /**
+   * Execute a modified version of try-and-increment, by using rehashing instead of simply
+   * incrementing the base point at each iteration.
+   * @param xCandidate base candidate for the x coordinate
+   * @return A random point on the curve determined from xCandidate
+   */
+  private ECPoint computePoint(BigInteger xCandidate) {
+    ECFieldElement x = this.curve.fromBigInteger(xCandidate);
+    ECFieldElement a = this.curve.getA();
+    ECFieldElement b = this.curve.getB();
+    ECFieldElement ySquare, y;
+    ECPoint resPoint;
+    do {
+      do {
+        x = nextFieldElement(x);
+        //      = x^3 + ax + b
+        ySquare = x.multiply(x.square()).add(a.multiply(x)).add(b);
+        y = ySquare.sqrt();
+      } while (y == null);
+      resPoint = this.curve.createPoint(x.toBigInteger(), y.toBigInteger());
+      // Ensure we have a point in the correct subgroup
+      resPoint = resPoint.multiply(this.curve.getCofactor()).normalize();
+      // Note that we need to validate the point. While multiplying with the co-factor should be
+      // sufficient.
+    } while (!resPoint.isValid());
+    return resPoint;
+  }
+
+  /**
+   * Hash the point to get a new random, but deterministic point on the curve
+   * this improves over the classical try-and-increment approach in that it will be random
+   * and not biased in case that the spread of valid x points are not uniform
+   * @param input the input to derive next integer from
+   * @return a new random integer
+   */
+  ECFieldElement nextFieldElement(ECFieldElement input) {
+    HmacDrbg rand = new HmacDrbg(input.getEncoded());
+    // The amount of bytes we need is bounded by the field size (in bites),
+    // but we take the field size twice before we compute modulo to ensure the distribution is
+    // statistically indistinguishable from random with field size being the statistical security
+    // parameter
+    int bytesNeeded = (2 * curve.getFieldSize()+7) / 8;
+    byte[] nextInt = new byte[bytesNeeded];
+    rand.nextBytes(nextInt);
+    BigInteger rawInt = new BigInteger(1, nextInt);
+    return curve.fromBigInteger(rawInt.mod(curve.getField().getCharacteristic()));
+  }
+
+  @Override
+  BouncyCastleECCElement decodeElement(byte[] bytes) {
+    return new BouncyCastleECCElement(this.curve.decodePoint(bytes));
```

Comment:
Ensure that the element is of the correct order (when the curve does not have cofactor 1). I.e. that the point is actually valid and that the point multiplied by the cofactor is different from the point at infinity.
It might also be relevant to check that it is not on the point of infinity.

Reply:
This is also an issue in `ECChouOrlandi` when decoding

Reply:
BouncyCastle already validates points when decoding, both that they are on the curve and that the order time.
It runs the following:
```java
    protected boolean satisfiesOrder()
    {
        if (ECConstants.ONE.equals(curve.getCofactor()))
        {
            return true;
        }

        BigInteger n = curve.getOrder();

        // TODO Require order to be available for all curves

        return n == null || ECAlgorithms.referenceMultiply(this, n).isInfinity();
    }
```
from `org.bouncycastle.math.ec.ECPoint`

Reply:
As mentioned above ideally manually check this as well or write a comment that it is expected from the library. But the functionality in the library might change, hence it is better to also check it manually since this can _completely_ unmine the security if not done

Reply:
Added a comment in 0375fd3